operators/src/engine/operator.rs

```diff
@@ -65,21 +43,236 @@ pub trait VectorOperator: Operator + CloneableVectorOperator {
+    fn initialized_operator(
+        self: Box<Self>,
+        context: ExecutionContext,
+    ) -> Result<Box<dyn InitializedRasterOperator>>;
+
+    /// Wrap a box around a `RasterOperator`
+    fn boxed(self) -> Box<dyn RasterOperator>
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+}
+
+pub type ExecutionContext = u32;
+
+pub trait InitializedOperator {
+    fn execution_context(&self) -> &ExecutionContext;
+
+    /// Get the sources of the `Operator`
+    fn raster_sources(&self) -> &[Box<dyn InitializedRasterOperator>];
+
+    /// Get the sources of the `Operator`
+    fn vector_sources(&self) -> &[Box<dyn InitializedVectorOperator>];
+
+    /// Get the sources of the `Operator`
+    fn raster_sources_mut(&mut self) -> &mut [Box<dyn InitializedRasterOperator>];
+
+    /// Get the sources of the `Operator`
+    fn vector_sources_mut(&mut self) -> &mut [Box<dyn InitializedVectorOperator>];
+}
+
+pub trait InitializedVectorOperator: Send + Sync {
     /// Get the result type of the `Operator`
-    fn result_type(&self) -> RasterDataType;
+    fn result_descriptor(&self) -> VectorResultDescriptor;
+
+    /// Instantiate a `TypedVectorQueryProcessor` from a `RasterOperator`
+    fn vector_processor(&self) -> Result<TypedVectorQueryProcessor>;
+
+    /// Wrap a box around a `RasterOperator`
+    fn boxed(self) -> Box<dyn InitializedVectorOperator>
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+}
+pub trait InitializedRasterOperator: Send + Sync {
+    /// Get the result type of the `Operator`
+    fn result_descriptor(&self) -> RasterResultDescriptor;
 
     /// Instantiate a `TypedRasterQueryProcessor` from a `RasterOperator`
     fn raster_processor(&self) -> Result<TypedRasterQueryProcessor>;
 
     /// Wrap a box around a `RasterOperator`
-    fn boxed(self) -> Box<dyn RasterOperator>
+    fn boxed(self) -> Box<dyn InitializedRasterOperator>
     where
         Self: Sized + 'static,
     {
         Box::new(self)
     }
 }
 
+impl InitializedRasterOperator for Box<dyn InitializedRasterOperator> {
+    fn result_descriptor(&self) -> RasterResultDescriptor {
+        self.as_ref().result_descriptor()
+    }
+    fn raster_processor(&self) -> Result<TypedRasterQueryProcessor> {
+        self.as_ref().raster_processor()
+    }
+}
+
+impl InitializedVectorOperator for Box<dyn InitializedVectorOperator> {
+    fn result_descriptor(&self) -> VectorResultDescriptor {
+        self.as_ref().result_descriptor()
+    }
+    fn vector_processor(&self) -> Result<TypedVectorQueryProcessor> {
+        self.as_ref().vector_processor()
+    }
+}
+
+pub struct InitilaizedOperatorImpl<Parameters, ResultDescriptor> {
+    pub params: Parameters,
+    pub raster_sources: Vec<Box<dyn InitializedRasterOperator>>,
+    pub vector_sources: Vec<Box<dyn InitializedVectorOperator>>,
+    pub context: ExecutionContext,
+    pub result_descriptor: ResultDescriptor,
```

Comment:
Why is it not necessary to have a common trait for the result description or sufficient to just use one struct?

Reply:
The ResultDescriptor is (currently) specific to the kind of operator. A VectorOperator has a VectorResultDescriptor and a RasterOperator a RasterResultDescriptor. We could add a common trait for things like projection.

Reply:
I guess that trait makes sense to group common functionality, which is for now the projection.